Fixes #34997 - Fix expand/collapse all behavior in Details tab

jeremylenz opened the pull request:
This is a refactor of the Expand/Collapse All behavior on the Details tab.

#### What hasn't changed:
* You can click to expand or collapse any individual card.
* Individual card states are saved in the browser's localStorage and loaded on page/tab load.

#### Before:
* There was an 'Expand / Collapse All' button which expands or collapses all cards.
* This state was kept in `isExpandedGlobal`, a React state independent of all individual card expansion states
* Clicking on 'Expand / Collapse All' would sometimes expand all and sometimes collapse all, depending on `isExpandedGlobal`
* Because of this, it was possible to get into states where 'Expand / Collapse All' does nothing, or does the opposite of what you expect

![expand-all](https://user-images.githubusercontent.com/22042343/189203997-102c3736-3688-4513-8d0c-5f0dbd2e0861.png)


#### After:
* The button text now changes as needed: When all cards are expanded, the button will be 'Collapse all cards.' Otherwise, it will be 'Expand all cards.' This behavior is much more clear and predictable.
* To accomplish this, the Details tab now knows the state of all its cards.
* I used React Context and useReducer to give the Details tab access to all card expansion states, and pass them directly to each card. This has several advantages: it avoids prop drilling through multiple components, and also through multiple code bases. The change is all in Foreman; I didn't have to change anything in Katello or other plugins. For more information on this pattern, see https://reactjs.org/docs/hooks-faq.html#how-to-avoid-passing-callbacks-down
* ~~In order for the Details tab to access all the card data, I load all the card names from the fill IDs in use for the slot.  To keep everything consistent I needed to have a consistent card ID string in both the fill ID and the card header. So I changed some fill IDs and added a (hopefully) helpful error message.~~
* The `isExpandedGlobal` prop is eliminated and you no longer need to pass it down in CardTemplate. Instead, `areAllCardsExpanded` is a computed value derived from all the individual card states.
* Instead of writing to localStorage on every state change, it now only writes on unmount / cleanup and reads on page / tab load
* This should work out of the box for any card that uses CardTemplate.

jeremylenz committed:
Fixes #34997 - Fix expand/collapse all behavior in Details tab

jeremylenz committed:
Refs #34997 - use fill IDs as card IDs

jeremylenz committed:
Refs #34997 - remove isExpandedGlobal from System Properties card

MariaAga commented on webpack/assets/javascripts/react_app/components/HostDetails/Tabs/Details/index.js:
```diff
@@ -1,12 +1,56 @@
+
+export const CardExpansionContext = React.createContext({});
+
+function cardExpansionReducer(state, action) {
```
Can this be an arrow function for consistency?

MariaAga commented on webpack/assets/javascripts/react_app/components/HostDetails/Templates/CardItem/CardTemplate/index.js:
```diff
@@ -13,36 +13,40 @@ import {
+  const { cardExpandStates, dispatch, cardIds } = useContext(
+    CardExpansionContext
   );
+  const cardId = cardIds?.find(id => id.includes(header));
```
This doesn't work well for a translated foreman, even if the id is translated it still doesnt always fit with the header unless its the same string exactly

MariaAga commented on webpack/assets/javascripts/react_app/components/HostDetails/Tabs/Details/index.js:
```diff
@@ -1,12 +1,56 @@
+      return collapsedState;
+    }
+    default:
+      throw new Error();
```
Can the error have some text, like `invalid card expansion type: ${type}`?

MariaAga commented on webpack/assets/javascripts/react_app/components/HostDetails/Tabs/Details/index.js:
```diff
@@ -16,18 +60,80 @@ const DetailsTab = ({ response, status, hostName }) => {
+
+  // On mount, get values from localStorage and set them in state
+  useEffect(() => {
+    if (cardIds?.length && cardIds.length !== cardCount.current) {
+      cardIds.forEach(key => {
+        const value = localStorage.getItem(`${key} card expanded`);
+        if (value !== null && value !== undefined) {
+          dispatch({ type: value === 'true' ? 'expand' : 'collapse', key });
+        } else {
+          dispatch({ type: 'add', key });
+        }
+      });
```
do we need this and `getInitialState`? they look very similar

MariaAga reviewed (changes requested):
Thanks for this fix! tested on foreman+plugins and it works well on untranslated foreman :)
needs to be changed so it will also work on translated foreman

jeremylenz committed:
Refs #34997 - don't use fill IDs

jeremylenz committed:
Refs #34997 - address comments

jeremylenz commented:
thanks @MariaAga!

Turns out using the fill IDs simply isn't compatible with translation at all. So I reworked it to just use the card headers, and ended up simplifying things quite a bit. I think I've also addressed the remaining comments. let me know..

jeremylenz commented:
I set my locale to German (hello Betriebssystem card!) and it seems to work fine now.

MariaAga reviewed (approved):
Thanks! everything works well now :)
